API rewritten to FastAPI

kamil-certat opened the pull request:
The FastAPI-based API is now an ASGI application, that isn't
natively supported by Apache. Instead, a recommended FastAPI
setup with web server Gunicorn is used. To keep backward compatibility,
Apache is still used as a Proxy. Apache and Gunicorn communicate
using unix socket.

The package installation creates intelmq-api service and socket
for systemd, and configure the Apache2 to communicate with them.
The INTELMQ_API_CONFIG env can now be set in the service unit file.

Provided configuration is an MVP and should be reviewed in the
production environement.

Depends on #38

kamil-certat committed:
DEV: added .vscode to the .gitignore

kamil-certat committed:
Dropped Python 3.6. Checked code style. New deps

As an intro for transition to the FastAPI and to be in
sync with the IntelMQ 3.1, the Python 3.6 was dropped.
In addition, the code style checks was introduced to the
CI. The config is the same as for IntelMQ core.

Dependencies for FastAPI was also added to check CI compatibility.

kamil-certat committed:
ENH: Initialy rewritten to FastAPI

kamil-certat committed:
ENH: Reorganized file structure

kamil-certat committed:
DOC: Updated readme and docstrings.

Readme contains now the updated description and new
development tips.

In addition, duplication of the CLI command for adding
user was removed. Small formatting changes was done.

kamil-certat committed:
FIX: Install libraries for type checks

kamil-certat committed:
FIX: Allow methods used by API

kamil-certat committed:
FIX: Fixed response serialization

kamil-certat committed:
ENH: Reduced rewriting of the CLI responses

kamil-certat committed:
TESTS: Install dependencies in CI. Add license

kamil-certat committed:
FIX: Fixed input for /api/run

kamil-certat committed:
FIX: Added compadibility paths, test saving config

IntelMQ Manager can produce invalid paths with doubled slashes,
and doesn't respect redirection. To keep backward compatibility,
affected paths are added in the format produced by Manager.

In addition, previous attempt with redirecting invalid paths
is removed, and saving runtime and position configurations
is now tested.

kamil-certat committed:
FIX: fixed typing

kamil-certat committed:
ENH: Package updated to setup new API

The FastAPI-based API is now an ASGI application, that isn't
natively supported by Apache. Instead, a recommended FastAPI
setup with web server Gunicorn is used. To keep backward compatibility,
Apache is still used as a Proxy. Apache and Gunicorn communicate
using unix socket.

The package installation creates intelmq-api service and socket
for systemd, and configure the Apache2 to communicate with them.
The INTELMQ_API_CONFIG env can now be set in the service unit file.

Provided configuration is an MVP and should be reviewed in the
production environement.

kamil-certat committed:
FIX: Disable building for Debian Buster

Dependency packages exist in the Debian Bullseye repositories, but
not in the Buster. For packaging Buster, we may need to package
dependencies

kamil-certat commented:
The changes in compare to the #38 are visible here: https://github.com/kamil-certat/intelmq-api/pull/1

This PR is ready to review, but will be opened after merging the #38

aaronkaplan commented on .github/workflows/code_style.yml:
```diff
@@ -0,0 +1,38 @@
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: 3.8
```
what about python 3.9, 3.10, 3.11?

kamil-certat replied:
We could do it, but please see this is just a style check, no unit tests run. I don't think it makes sense to run style check on every Python version. It may sometimes differ, but should not.

sebix replied:
I agree with @kamil-certat

aaronkaplan replied:
then we leave it.

aaronkaplan commented on .github/workflows/mypy.yml:
```diff
@@ -18,7 +18,7 @@ jobs:
       fail-fast: false
       matrix:
-        python-version: [3.6, 3.7, 3.8]
+        python-version: [3.7, 3.8]
```
same... 3.9, 3.10, 3.11?

kamil-certat replied:
The same answer - I don't really think we need to run type check on every Python version. Let me know if you think it makes sense, then I'll do it.

aaronkaplan replied:
convinced

aaronkaplan commented on .github/workflows/python-unittests.yml:
```diff
@@ -18,7 +18,7 @@ jobs:
       fail-fast: false
       matrix:
-        python-version: [3.6, 3.7, 3.8]
+        python-version: [3.7, 3.8]
```
same

sebix replied:
Here I agree with @aaronkaplan to keep it in sync with the supported and tested platforms of intelmq (up to 3.11)

kamil-certat replied:
Yeah, I've missed that place. Done.

aaronkaplan commented on contrib/intelmq-api.service:
```diff
@@ -0,0 +1,26 @@
+User=www-data
+Group=www-data
+RuntimeDirectory=gunicorn
+WorkingDirectory=/usr/lib/python3/dist-packages/intelmq_api/
```
I might suggest to change the path  into a variable... why? different t linux distribs

sebix replied:
The file is an example, and this path works on most distros. At least on all supported ones (Debian and Ubuntu, that is).

kamil-certat replied:
This is a good point, but I'm unsure how to solve it at the moment. The enviromental variable doesn't really makes sense - it should still need to be overriden by user, what is the same as override the WorkingDirectory directive... Do you have any ideas? Maybe the installation script should detect the platform?

aaronkaplan replied:
nope, then we leave it as is

aaronkaplan commented on contrib/intelmq-api.socket:
```diff
@@ -0,0 +1,12 @@
+Description=The socket to handle IntelMQ API requests
+
+[Socket]
+ListenStream=/usr/lib/python3/dist-packages/intelmq_api/intelmq_api.sock
```
Same here. the path might be distrib dependent. We also have Redhat users for example.

aaronkaplan replied:
leaving it since it's an example in contrib/

aaronkaplan commented on debian/changelog:
```diff
@@ -1,3 +1,11 @@
+intelmq-api (3.1.0~rc2) UNRELEASED; urgency=medium
+
+  * API rewritten to use the FastAPI framework
```
\o/

aaronkaplan commented on intelmq_api/main.py:
```diff
@@ -0,0 +1,31 @@
+
+@app.on_event("startup")
+def init_app():
+    intelmq_api.dependencies.startup(config)
```
just wondering why you use dependencies here... seems like extra complicated (-> KISS principle). Wouldn't it be enough to just call the loading of the config once here (I mean directly, without doing a dependency)?

kamil-certat replied:
I'm unsure if I understand you correctly.

This hook is calling a function from our module 'dependencies', that does exactly what you said - load config once, and save it in FastAPI-style dependencies. Then, FastAPI-dependencies are used to inject configuration or session store when processing requests. The one-time loading on startup is done to follow the previous solutions with globals, but in a clean way. Thanks to using dependencies, we can easily inject prepared configuration in tests.

aaronkaplan replied:
its ok. I was just wondering.

aaronkaplan commented on intelmq_api/main.py:
```diff
@@ -0,0 +1,31 @@
+
+app.add_middleware(CORSMiddleware, allow_origins=config.allow_origins,
+                   allow_methods=("GET", "POST"))
+app.include_router(api, prefix="/v1")
```
Are we compatible with the old API endpoint? I mean - should't we intentionally increase the API version number here?

kamil-certat replied:
My idea and work was to be fully compatible

aaronkaplan commented on intelmq_api/intelmq-api.wsgi:
```diff
@@ -3,6 +3,9 @@
 SPDX-FileCopyrightText: 2020 Birger Schacht
 SPDX-License-Identifier: AGPL-3.0-or-later
 """
+
```
Is this file still needed? Or the TODO here?

kamil-certat replied:
Ahh, leftover

aaronkaplan commented on scripts/run_dev.sh:
```diff
@@ -0,0 +1,4 @@
+# SPDX-FileCopyrightText: 2022 CERT.at GmbH <https://cert.at>
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+uvicorn intelmq_api.main:app --reload
```
--reload needed? I guess that depends on how we run it actually. In Docker, then yes probably...

kamil-certat replied:
No, it's not nedded in Docker (why should be?), it's needed in development environement. The --reload keeps track on changed files and autmatically reload when something changed. It's just to help developers working on the API, and this script is just for developers.

aaronkaplan commented on intelmq_api/version.py:
```diff
@@ -3,5 +3,5 @@
 SPDX-License-Identifier: AGPL-3.0-or-later
 """
-__version_info__ = (3, 1, 0)
+__version_info__ = (3, 1, 0, 'rc2')
```
will need to be adapted for the final 3.1 rel. XXX TODO XXX

kamil-certat replied:
Changed.

aaronkaplan commented on intelmq_api/api.py:
```diff
@@ -12,194 +12,162 @@
-@typing.no_type_check
-def botnet(action: Actions, group: Groups = None):
-    return runner.botnet(action, group)
+@api.get("/api/botnet", dependencies=[authorized])
```
isn't the new path /api/v1/... ?

kamil-certat replied:
No, it's /v1/api/... - and the /v1/ is set in main.py

aaronkaplan commented on intelmq_api/dependencies.py:
```diff
@@ -0,0 +1,59 @@
+T = TypeVar("T")
+
+
+class OneTimeDependency(Generic[T]):
```
Do we need this type of dependency ? Or does it merely add lines of code.
KISS principle.

kamil-certat replied:
If we don't, then they will be loaded more than one time (on every request?). It could work, however I took a conservative approach to emulate how it was warking previously: this is exactly how we had it. There were no tests I can use to ensure that moving to typical dependencies doesn't break anything (like parallel access to sqllite etc.)

aaronkaplan replied:
I see, thanks for the explanations.

aaronkaplan commented on intelmq_api/api.py:
```diff
@@ -12,194 +12,162 @@
-def initialize_api(config: intelmq_api.config.Config) -> None:
-    """Initialize the API, optionally loading a configuration file.
+def ID(id: str) -> str:
+    if not set(id) < ID_CHARS:
```
elegant :)

kamil-certat replied:
thanks :)

aaronkaplan commented on intelmq_api/api.py:
```diff
@@ -12,194 +12,162 @@
-    """
-    return hug.get(*args, **kw).cache(max_age=3)
 
+class LoginForm(BaseModel):
```
(mior change request): I propose that we move all the fastAPI models to a separate python file.

kamil-certat replied:
Definitely good move!

aaronkaplan commented:
@kamil-certat I reviewed the PR. Could you please have a look at my comments, then I will be very happy to merge it in and test.
My main point - apart from the lots of little comments - is: did you test the old API vs. new API code?
Like are there any diffs when given the same input?

input -> old api -> output1
input -> new api -> output2

diff output1 output2 

?

kamil-certat commented:
@aaronkaplan I don't see any your comments, have you published them? The diff question answered in #38

aaronkaplan commented:
> @aaronkaplan I don't see any your comments, have you published them? The diff question answered in #38

Hm, yeah... you should see all my comments in the PR on github. Ping me in case you don't

kamil-certat commented:
I still don't see anything to address...

kamil-certat committed:
FIX: Use previous HTTP code&msg for rejected auth

Previously API was using a 401 and specific message when
rejecting auth. This should be kept to ensure backward
compatibility

aaronkaplan commented:
now?

kamil-certat commented:
Now I see, will answer shortly :)

sebix commented:
> @[aaronkaplan](https://github.com/aaronkaplan) aaronkaplan requested a review from [sebix](https://github.com/sebix)

Sorry, but I can't manage to review the API rewrite until Friday. That would include tests on various platforms, verifying documentation and examples and updating the package build instructions for multiple operating systems.

kamil-certat commented:
As I mentioned in some e-mail, there is trouble that for Debian < 11 packages for FastAPI don't exist. I think it would be a terrible exercise to prepare all of them in two days.

sebix commented on contrib/intelmq-api.socket:
```diff
@@ -0,0 +1,12 @@
+SocketUser=www-data
+
+[Install]
+WantedBy=sockets.target
```
missing newline

sebix commented on debian/install:
```diff
@@ -1,3 +1,5 @@
 contrib/api-sudoers.conf => /etc/sudoers.d/01_intelmq-api
 contrib/api-session.sql => /usr/share/dbconfig-common/data/intelmq-api/install/sqlite3
+contrib/intelmq-api.service => /lib/systemd/system/intelmq-api.service
+contrib/intelmq-api.socket => /lib/systemd/system/intelmq-api.socket
```
missing newline

sebix commented on debian/install:
```diff
@@ -1,3 +1,5 @@
 #!/usr/bin/dh-exec
 contrib/api-sudoers.conf => /etc/sudoers.d/01_intelmq-api
 contrib/api-session.sql => /usr/share/dbconfig-common/data/intelmq-api/install/sqlite3
+contrib/intelmq-api.service => /lib/systemd/system/intelmq-api.service
```
Are you sure that `/lib/systemd/` is the correct path?

kamil-certat replied:
For Debian, but not for RedHat... https://unix.stackexchange.com/a/458252

sebix commented on debian/postinst:
```diff
@@ -3,11 +3,13 @@
 
         #adduser www-data intelmq
-	usermod -a -G intelmq www-data
+		usermod -a -G intelmq www-data
```
strange indent

kamil-certat replied:
No, let's open the file. It was strange previously ;)

sebix commented on contrib/api-apache.conf:
```diff
@@ -1,16 +1,15 @@
-   Require all granted
-</Directory>
+
+<VirtualHost *:80>
```
This file is/was not a vhost definition, in other words, it's not placed in `/etc/apache2/sites-available/`, but in `/etc/apache2/conf-available/` (Debian-style paths), independent of the vhost definition and thus independent of any HTTPS vhost definition.

kamil-certat replied:
Fixed, thanks!

sebix commented on contrib/intelmq-api.service:
```diff
@@ -0,0 +1,26 @@
+PrivateTmp=true
+
+[Install]
+WantedBy=multi-user.target
```
missing newline

sebix reviewed (changes requested):
Did you verify that the packaging works on the supported platforms?

kamil-certat commented on contrib/api-apache.conf:
```diff
@@ -1,16 +1,15 @@
+	</Location>
+
+   <Directory /usr/lib/python3/dist-packages/intelmq_api/>
+      Require all granted
```
TODO: Ensure we still need it.

kamil-certat replied:
We don't.

kamil-certat committed:
ENH: Improvements from review

kamil-certat commented:
@sebix Unfortunately not, and this is why I wanted to separate rewrite and packaging at the beginning. It was, in fact, the first approach to package it.

It works on Debian 11, on the Debian 10 - as mentioned - we have lack of dependencies. RedHat's platform need to be tested.

kamil-certat committed:
FIX: Force yaml not to convert to number

kamil-certat committed:
ENH: Remove unused Apache directive

sebix commented on .gitignore:
```diff
@@ -6,3 +6,5 @@ build
 *.egg-info
 *.sqlite
+
+.vscode/
```
missing newline

sebix commented on CHANGELOG.rst:
```diff
@@ -1,10 +1,39 @@
+
+**Known issues:**
+
+- The packaging isn't yet prepared to work with FastAPI-based API.
```
Is this true? This PR changes a few files related to packaging

kamil-certat replied:
Ugh, I missed this. I think my first idea was to separately handle API rewriting and packaging, bc I was expecting more troubles with packages than the API. This was written in the first PR.

aaronkaplan replied:
so, is that fixed now or not?

aaronkaplan commented:
> > @[aaronkaplan](https://github.com/aaronkaplan) aaronkaplan requested a review from [sebix](https://github.com/sebix)
> 
> Sorry, but I can't manage to review the API rewrite until Friday. That would include tests on various platforms, verifying documentation and examples and updating the package build instructions for multiple operating systems.

okay, got it. @kamil-certat would it be OK to put that into 3.2 then?

kamil-certat commented:
I'm okay with this or just releasing API 3.1 a little later this week. Let me do additional tests and packaging fixes, hopefully today or tomorrow. If I fail with proper packaging until tomorrow, I may need help here if we want to finish this week.

And any additional testing is welcome :)

kamil-certat commented:
Anyway, I will not back-port dependencies for Debian 10 (or eventually CentOS, if needed) - will you eventually handle it during release packaging, @sebix?

sebix commented:
> Anyway, I will not back-port dependencies for Debian 10 (or eventually CentOS, if needed) - will you eventually handle it during release packaging, @sebix?

If there's no explicit contract or funding: No.

kamil-certat commented:
Sure, I understand it. Then we have a releasing issue because it works out-of-the-box for Debian 11, but will fail for the earlier.

@aaronkaplan: this is a question to you, how you think it's the best to handle it. I see the following options:
1) we can wait until someone find time to back-port dependencies (or maybe someone has already done it somewhere);
2) let's release API 3.1 with native packages for Debian 11, and leave systems without already built FastAPI on the PyPI packages as newest and 3.0.1 as native package (should work so far); change it once we find time to back-port dependencies. Already done native support doesn't look good: https://pkgs.org/search/?q=fastapi
3) abandon the FastAPI and look for a framework that is already widely supported; flask looks [promising](https://pkgs.org/search/?q=flask) and could be served as WSGI, so the same way as hug, by the Apache.

@sebix - is it possible to contribute to your building server?

sebix commented:
I don't think backporting fastapi is useful. It has some dependencies which are not fulfilled in Debian 10, all of them need backports as well.

> @sebix - is it possible to contribute to your building server?

It's not _mine_, but a publicly usable instance. Contact me privately for the details (username etc.)

kamil-certat commented:
> I don't think backporting fastapi is useful. It has some dependencies which are not fulfilled in Debian 10, all of them need backports as well.

Yeah, it would need the full back-porting, the same way as we do with hug know. I'm also not a fan of maintaining all package dependencies on our own. So, what would you suggest? 

> t's not mine, but a publicly usable instance. Contact me privately for the details (username etc.)

I know, I called it yours because the repo is under your namespace (sebix/intelmq) ;) I'll contact you once I find it useful to change something :)

kamil-certat commented:
Okay, so after the mailing thread, my goals for packaging are:

1) native package for Debian 11 (all dependencies exists),
2) for other platforms, the pip package as recommended way.

To achieve 2) I'll update the documentation (PR https://github.com/certtools/intelmq/pull/2292) and try to prepare a script that should help with installing service units. In the future, it could be somehow integrated with `intelmqsetup` (but not at the moment, to not introduce additional complexity).

aaronkaplan commented:
> As I mentioned in some e-mail, there is trouble that for Debian < 11 packages for FastAPI don't exist. I think it would be a terrible exercise to prepare all of them in two days.

Yes, okay. Makes sense. So, what I propose is that we add a sentence to the documentation (if not already done) that only versions x,y,z are supported (Debian 11, Ubuntu  22, etc.) and what steps users of older OSes or other distros can do to get it running. OK?

kamil-certat commented:
Exactly

kamil-certat committed:
Imporved comments and updated changelod

kamil-certat committed:
ENH: Add script to setup Systemd units

kamil-certat committed:
FIX: Fixed installing unit paths

kamil-certat committed:
Missed newline

kamil-certat commented:
Please take a look now. I'll follow up tomorrow with final testing, today I didn't manage to finally test CentOS (some VirtualBox issues).

kamil-certat committed:
Update changelog with setup script

kamil-certat committed:
Cleaner changing values

sebix commented on contrib/api-apache.conf:
```diff
@@ -1,16 +1,9 @@
-   Require all granted
-</Directory>
+
+# If you want to change default location, please allign the ROOT_PATH in the service configuration
```
typo "allign"

kamil-certat committed:
Fix typo

kamil-certat committed:
Fix Debian changelog

kamil-certat committed:
FIX: Login endpoint compatibility

It was previously missed, that the login endpoint unintentionally
changes the input format from HTTP Form to JSON. It's now fixed
and tested.

kamil-certat committed:
FIX: Fix compatibility with FastAPI < 0.78.0

Prior to FastAPI 0.78.0, the Form() object required explicit
setting with eclipse when a form field is required. As the
FastAPI in Debian 11 repository is older, we need to support it

kamil-certat committed:
Update minimal fastapi requirements

sebix commented on debian/control:
```diff
@@ -3,11 +3,14 @@ Section: python
 Priority: optional
 Maintainer: IntelMQ Team <intelmq-team@cert.at>
 Uploaders: Birger Schacht <schacht@cert.at>
```
this line needs a change too

kamil-certat replied:
Good catch, but who should be there? I assume that you are the one, who looks the best as `Uploader`, aren't you?

sebix replied:
ack, makes sense

sebix reviewed (changes requested):
These packaged files are in the wrong location:
* `/etc/intelmq/intelmq-api.service`
* `/etc/intelmq/intelmq-api.socket`
* `/etc/intelmq/api-sudoers.conf`

The tests are not executed at build time:
```
[   56s]    dh_auto_test -O--buildsystem=pybuild
[   56s] I: pybuild base:232: cd /usr/src/packages/BUILD/.pybuild/cpython3_3.9/build; python3.9 -m unittest discover -v 
[   56s] 
[   56s] ----------------------------------------------------------------------
[   56s] Ran 0 tests in 0.000s
[   56s] 
[   56s] OK
```

kamil-certat commented:
Could you please elaborate, what do you mean by `wrong place`? Files you mentioned are in `/etc/intelmq` because of the `setup.py` configuration. The `/etc/intelmq/api-sudoers.conf` was always there. Do you mean, that we shouldn't include them in the python package, or in another dir? The Debian config placed them in the right dirs separately:

https://github.com/kamil-certat/intelmq-api/blob/727c3563cb79e72e770213ed739864e7ac683146/debian/install#L1-L5

kamil-certat commented:
re: tests. They were never running ;) https://build.opensuse.org/package/live_build_log/home:sebix:intelmq/intelmq-api/Debian_10/x86_64

We could make them working, but it would require downloading the IntelMQ package as Build-Depends in the build server. Do you know how to configure OBS to get package from another project?

sebix commented:
> Do you know how to configure OBS to get package from another project?

it will pick the intelmq package from the same repo (i.e. unstable). Why do you need another one? OBS uses the packages in the same project and then the repositories as defined per distributions (with preference on the local packages). I don't see a reason why you need to specify the project for your purpose.

> re: tests. They were never running ;) https://build.opensuse.org/package/live_build_log/home:sebix:intelmq/intelmq-api/Debian_10/x86_64

now we have better tests (previously only the session store was tested) and it makes more sense now

sebix commented:
> The `/etc/intelmq/api-sudoers.conf` was always there. Do you mean, that we shouldn't include them in the python package, or in another dir? The Debian config placed them in the right dirs separately:
> 
> https://github.com/kamil-certat/intelmq-api/blob/727c3563cb79e72e770213ed739864e7ac683146/debian/install#L1-L5

The Debian package does place them correctly but does not remove the files placed incorrectly by setuptools. To cut a long story short, I recommend (as the simplest and explicit solution) removing the wrongly placed files in the `debian/rules` file, section `override_dh_auto_install`:
```
rm debian/intelmq-api/etc/....
```

kamil-certat commented:
> it will pick the intelmq package from the same repo (i.e. unstable). Why do you need another one? OBS uses the packages in the same project and then the repositories as defined per distributions (with preference on the local packages). I don't see a reason why you need to specify the project for your purpose.

Okay, I'll try it :)

kamil-certat committed:
FIX: Updater name. Remove config files from wrong places

kamil-certat commented:
I fixed what you asked for @sebix 

However, building for Ubuntu 22.04 is broken due to the upstram bug: https://bugs.launchpad.net/ubuntu/+source/fastapi/+bug/1970557 Basically, the Ubuntu FastAPI and Starlette packages are out-of-sync in this version, and FastAPI is unable to work.

sebix commented:
> I fixed what you asked for @sebix

Thanks, let's merge then!

> However, building for Ubuntu 22.04 is broken due to the upstram bug: https://bugs.launchpad.net/ubuntu/+source/fastapi/+bug/1970557 Basically, the Ubuntu FastAPI and Starlette packages are out-of-sync in this version, and FastAPI is unable to work.

https://bugs.launchpad.net/ubuntu/+source/fastapi/+bug/1970557/comments/6 suggests that backporting 0.74.1-1 from Ubuntu 22.10 to Ubuntu 22.04 would work without requiring any updates of depending packages. Would this be a viable solution?
